### ISSUE TITLE
Model building function arguments not clearly documented

### ISSUE THREAD

**Issue by stsievert** [adriangb/scikeras#57]
The arguments for `build_fn` are not clearly documented. For example, it's not clear where the *required* arguments below come from: 

https://github.com/adriangb/scikeras/blob/cf7ea78d062fbd5a2b9201b402422cce4afa95d7/tests/mlp_models.py#L8-L11

I don't see any place the user specified the number of features or number of outputs. This is especially confusing because the user typically knows the required arguments ahead of time because they have a dataset and problem in mind.

The docstring for `BaseWrapper._fit_build_keras_model` says the following:

https://github.com/adriangb/scikeras/blob/cf7ea78d062fbd5a2b9201b402422cce4afa95d7/scikeras/wrappers.py#L155-L156

How the "appropriate arguments" determined?

**Comment by stsievert** [comment]
I think the following would resolve this issue:

1. Allow the user to control which arguments get passed to `build_fn` via passing `model__n_features`, `model__n_classes` to `KerasClassifier.__init__`, then piping them to `build_fn`.
2. Clearly documenting which parameters can be passed and where they come from.

The documentation for (1) would be a lot clearer and a lot shorter.

**Comment by adriangb** [comment]
> it's not clear where the _required_ arguments below come from:

Those are instance attributes of `BaseWrapper` set by `BaseWrapper.fit`. You are right that it is not clearly documented what parameters are available.

The way this works is that _any_ instance attribute that matches a parameter name in `build_fn` will be passed to it.

**Comment by adriangb** [comment]
> Allow the user to control which arguments get passed to `build_fn` via passing `model__n_features`, `model__n_classes` to `KerasClassifier.__init__`, then piping them to `build_fn`.

I envision something like this being part of the param_routing/parameter precedence effort. `build_fn` should be able to request `n_features_in_` (which is generated by `fit`) but the user should have the ability to specify `model__n_features_in_` in which case `build_fn(n_features_in_, ...)` would get the value of `model__n_features_in_` instead of `n_features_in_`

**Comment by stsievert** [comment]
> Do you think it might help to document what instance attributes are available for build_fn to request (i.e. n_features_in_, cls_type_, etc.)?

I think the simplest solution is parameter routing. The documentation would be simple: "Parameters passed to `BaseWrapper.__init__` prefixed with `model__` will be passed to the function `model`." This solution would also reduce the amount of code, at least 14 lines ([1], [2]).

[1]:https://github.com/adriangb/scikeras/blob/cf7ea78d062fbd5a2b9201b402422cce4afa95d7/scikeras/wrappers.py#L331-L340
[2]:https://github.com/adriangb/scikeras/blob/cf7ea78d062fbd5a2b9201b402422cce4afa95d7/scikeras/wrappers.py#L180-L185

I think the documentation around which instance attributed are available (solution (2)) would be longer, so it'd be harder to maintain. I think this solution would make SciKeras harder to use: the user would have to look at the documentation more frequently to determine what parameters are allowed.

**Comment by adriangb** [comment]
I'm confused as to how users would route parameters in __init__ that are created internally by `fit`.

**Comment by stsievert** [comment]
> I'm confused as to how users would route parameters in __init__ that are created internally by fit.

Why would that be required?

**Comment by adriangb** [comment]
There are many things that that enables, but two quick ones are:
- Set input dimensions based on `n_features_in_`
- Choose loss function for classifiers based on `cls_type_`.

**Comment by stsievert** [comment]
Those sound useful for testing. But I'm having a hard time imagining the user for the typically SciKeras user, someone that has the data and model on hand.

When would they find useful to use a parameter created internally by `fit`? And why can't they set the relevant parameters themselves? (e.g, `model__n_features=X.shape[1]` or `model__n_classes=len(np.unique(y))`).

**Comment by adriangb** [comment]
I think it is an important feature of this package to be able to create a self-contained estimator that is able to handle multiple input shapes and output types, like most Scikit-Learn estimators. That's actually the genesis of this entire package, it all started [here](https://github.com/tensorflow/tensorflow/issues/32532).

I don't think this should be hard to document, since we should be listing all of these attributes anyway.

I also think that parameter routing and parameter precedence can and should co-exist with the ability to pull in these instance attributes created by `fit`.

**Comment by stsievert** [comment]
> an important feature of this package to be able to create a self-contained estimator that is able to handle multiple input shapes and output types, like most Scikit-Learn estimators.

That's a better use case. I can see the motivation now.

I think it'd be cleaner to pass a keyword argument `meta` if it's accepted by the model building function, rather than determining some priority scheme. `meta` could be a dictionary with all the relevant parameters. I'm imagining this interface:

``` python
def model_fn(meta=None):
    assert set(meta.keys()) == {"n_features", "n_classes", "n_outputs", ...}
    ...
    return model
```

This would also enhance ease-of-use for the user: they can very quickly see which parameters are available by inspecting `meta.keys()`.

**Comment by adriangb** [comment]
That does look promising, I was already thinking of having a `meta_parameters` dict that is used to "overwrite" `__init__` params such that we don't have to use `_random_state` to overwrite `random_state` and instead can do:

```python3
def fit(X, y....):
    meta_params = self.get_params()  # initialize meta_params
    meta_params["random_state"] = ...whatever conversion is needed
    self._fit_keras_model(X, y,..., meta_params=meta_params)
```

Then we could easily filter these parameters and remove the routing prefixes before passing them onto their destination.
